From fa968234a50a362dec22e27548e424c781e58fed Mon Sep 17 00:00:00 2001
From: Lily Zhang <r58066@freescale.com>
Date: Mon, 15 Sep 2008 14:44:07 +0800
Subject: [PATCH] ENGR00091385 MC13892 touch screen can not work well

It was caused by signed and unsigned conversion. So the mask operation is added
into pmic_i2c_24bit_read function.

Signed-off-by: Lily Zhang <r58066@freescale.com>
---
 drivers/mxc/pmic/core/mc13892.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/drivers/mxc/pmic/core/mc13892.c b/drivers/mxc/pmic/core/mc13892.c
index b938785..195b961 100644
--- a/drivers/mxc/pmic/core/mc13892.c
+++ b/drivers/mxc/pmic/core/mc13892.c
@@ -50,7 +50,9 @@ int pmic_i2c_24bit_read(struct i2c_client *client, unsigned int reg_num)
 	ret = i2c_smbus_read_i2c_block_data(client, reg_num, 3, buf);
 
 	if (ret == 3) {
-		ret = buf[0] << 16 | buf[1] << 8 | buf[2];
+		ret =
+		    ((buf[0] << 16) & 0xFF0000) | ((buf[1] << 8) & 0xFF00) |
+		    (buf[2] & 0xFF);
 		return ret;
 	} else {
 		pr_debug("24bit read error, ret = %d\n", ret);
-- 
1.5.4.4

